drivers/i2c/i2c_ite_it8xxx2.c

```diff
@@ -0,0 +1,873 @@
+	{ &GPDMRE, &GPDMRE,	0x01, 0x80},
+	{ &GPDMRA, &GPDMRA,	0x10, 0x20},
+};
+
+static int i2c_get_line_levels(const struct device *dev)
+{
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+	uint8_t *base = config->base;
+	int pin_sts = 0;
+
+	if (config->port < I2C_STANDARD_PORT_COUNT)
+		return IT83XX_SMB_SMBPCTL(base) & 0x03;
+
+	if (*i2c_pin_regs[config->port].mirror_clk &
+					i2c_pin_regs[config->port].clk_mask)
+		pin_sts |= I2C_LINE_SCL_HIGH;
+	if (*i2c_pin_regs[config->port].mirror_data &
+					i2c_pin_regs[config->port].data_mask)
+		pin_sts |= I2C_LINE_SDA_HIGH;
+
+	return pin_sts;
+}
+
+static int i2c_is_busy(const struct device *dev)
+{
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+	uint8_t *base = config->base;
+
+	if (config->port < I2C_STANDARD_PORT_COUNT)
+		return (IT83XX_SMB_HOSTA(base) &
+					(HOSTA_HOBY | HOSTA_ALL_WC_BIT));
+
+	return (IT83XX_I2C_STR(base) & E_HOSTA_BB);
+}
+
+static void i2c_reset(const struct device *dev, int cause)
+{
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+	uint8_t *base = config->base;
+
+	if (config->port < I2C_STANDARD_PORT_COUNT) {
+		/* bit1, kill current transaction. */
+		IT83XX_SMB_HOCTL(base) = 0x2;
+		IT83XX_SMB_HOCTL(base) = 0;
+		/* W/C host status register */
+		IT83XX_SMB_HOSTA(base) = HOSTA_ALL_WC_BIT;
+	} else {
+		/* State reset and hardware reset */
+		IT83XX_I2C_CTR(base) = E_STS_AND_HW_RST;
+	}
+	printk("I2C ch%d reset cause %d\n", config->port, cause);
+}
+
+/*
+ * Set i2c standard port (A, B, or C) runs at 400kHz by using timing registers
+ * (offset 0h ~ 7h).
+ */
+static void i2c_standard_port_timing_regs_400khz(int port)
+{
+	/* Port clock frequency depends on setting of timing registers. */
+	IT83XX_SMB_SCLKTS(port) = 0;
+	/* Suggested setting of timing registers of 400kHz. */
+	IT83XX_SMB_4P7USL = 0x6;
+	IT83XX_SMB_4P0USL = 0;
+	IT83XX_SMB_300NS = 0x1;
+	IT83XX_SMB_250NS = 0x2;
+	IT83XX_SMB_45P3USL = 0x6a;
+	IT83XX_SMB_45P3USH = 0x1;
+	IT83XX_SMB_4P7A4P0H = 0;
+}
+
+/* Set clock frequency for i2c port A, B , or C */
+static void i2c_standard_port_set_frequency(const struct device *dev,
+					int freq_khz, int freq_set)
+{
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+
+	/*
+	 * If port's clock frequency is 400kHz, we use timing registers
+	 * for setting. So we can adjust tlow to meet timing.
+	 * The others use basic 50/100/1000 KHz setting.
+	 */
+	if (freq_khz == 400) {
+		i2c_standard_port_timing_regs_400khz(config->port);
+	} else {
+		IT83XX_SMB_SCLKTS(config->port) = freq_set;
+	}
+
+	/* This field defines the SMCLK0/1/2 clock/data low timeout. */
+	IT83XX_SMB_25MS = I2C_CLK_LOW_TIMEOUT;
+}
+
+/* Set clock frequency for i2c port D, E , or F */
+static void i2c_enhanced_port_set_frequency(const struct device *dev,
+				int freq_khz)
+{
+	struct i2c_it8xxx2_data *data = DEV_DATA(dev);
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+	uint8_t *base = config->base;
+	int clk_div, psr;
```

Comment:
Using signed values is probably wrong here, and could introduce undefined behavior with overflows.  These should be unsigned, and the integer constants in the calculations below have `U` suffices (e.g. `2U * 1000U ...) - 2U;`)

Reply:
Done